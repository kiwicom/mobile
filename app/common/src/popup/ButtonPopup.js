// @flow

import * as React from 'react';
import { View, StyleSheet } from 'react-native';

import Popup from './Popup';
import Button from '../buttons/Button';

type Props = {|
  buttonTitle: string,
  children: React.Node,
  isVisible: boolean,
  onClose: Function,
  onSave: Function,
  doScroll?: boolean,
  style?: Object,
|};

const POPUP_PADDING = 20;

/**
 * Simple popup with confirmation button at the bottom.
 */
export default class ButtonPopup extends React.Component<Props> {
  onClose = () => this.props.onClose();

  onSave = async () => {
    await this.props.onSave();
    this.onClose();
  };

  render = () => (
    <Popup
      isVisible={this.props.isVisible}
      onClose={this.onClose}
      doScroll={this.props.doScroll}
    >
      <View
        style={[styles.content, this.props.style && this.props.style.content]}
      >
        {this.props.children}
      </View>
      <Button
        title={this.props.buttonTitle}
        styles={{ button: styles.button }}
        onPress={this.onSave}
      />
    </Popup>
  );
}

const styles = StyleSheet.create({
  content: {
    padding: POPUP_PADDING,
    paddingBottom: 0,
  },
  button: {
    marginTop: 20,
    margin: POPUP_PADDING,
  },
});
