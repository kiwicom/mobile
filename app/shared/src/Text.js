// @flow

import * as React from 'react';
import ReactNative from 'react-native';

import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../types/Styles';

type Props = {|
  children: React.Node,
  style?: StylePropType,
|};

export default function Text({ children, style }: Props) {
  return (
    <ReactNative.Text style={[styles.nativeText, style]}>
      {children}
    </ReactNative.Text>
  );
}

const styles = StyleSheet.create({
  // These values are from the official design. Don't touch it please.
  nativeText: {
    fontWeight: 'normal',
    color: '#30363d',
    android: {
      fontSize: 15,
      lineHeight: 21,
    },
    ios: {
      fontSize: 14,
      lineHeight: 20,
    },
  },
});
