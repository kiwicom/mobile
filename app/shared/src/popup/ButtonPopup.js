// @flow

import * as React from 'react';
import { View, StyleSheet } from 'react-native';

import Popup from './Popup';
import Button from '../buttons/Button';

type Props = {|
  buttonTitle: string,
  children: React.Node,
  isVisible: boolean,
  onClose: Function,
  onSave: Function,
  style?: Object,
|};

const POPUP_PADDING = 20;

/**
 * Simple popup with confirmation button at the bottom.
 */
export default class ButtonPopup extends React.Component<Props> {
  onClose = () => this.props.onClose();

  onSave = () => this.props.onSave();

  render = () => (
    <Popup isVisible={this.props.isVisible} onClose={this.onClose}>
      <View
        style={[styles.content, this.props.style && this.props.style.content]}
      >
        {this.props.children}
      </View>
      <View style={styles.button}>
        <Button title={this.props.buttonTitle} onPress={this.onSave} />
      </View>
    </Popup>
  );
}

const styles = StyleSheet.create({
  content: {
    padding: POPUP_PADDING,
    paddingBottom: 0,
  },
  button: {
    marginTop: 20,
    margin: POPUP_PADDING,
  },
});
