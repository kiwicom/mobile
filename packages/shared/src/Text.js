// @flow

import * as React from 'react';
import PropTypes from 'prop-types';
import ReactNative from 'react-native';
import { defaultTokens } from '@kiwicom/mobile-orbit';

import StyleSheet from './PlatformStyleSheet';
import type { StylePropType } from '../types/Styles';

// basically every component (but never strings - they must be translated)
// the idea is that you should always use component as a children so you'll
// eventually end up with the Translation component
type ChildrenType = React.Element<any>; // FIXME: we cannot accept View in the children because it's not supported on Android

type Props = {|
  +children: ChildrenType | React.ChildrenArray<ChildrenType>,
  +style?: StylePropType,
  +numberOfLines?: number,

  // none - View (Text in this case) is never the target of touch events
  +pointerEvents?: 'none',
|};

/**
 * React Native has the concept of style inheritance, but limited to text
 * subtrees. In this case, the second part will be both bold and red:
 *
 * ```
 * <Text style={{fontWeight: 'bold'}}>
 *   I am bold
 *   <Text style={{color: 'red'}}>and red</Text>
 * </Text>
 * ```
 *
 * It's roughly translated to this (`NSAttributedString` or `SpannableString`):
 *
 * ```
 * "I am bold and red"
 * 0-9: bold
 * 9-17: bold, red
 * ```
 *
 * This style inheritance works for deeply nested 'Text' components as well.
 *
 * Known issues:
 * @see https://github.com/facebook/react-native/issues/10712
 */
export default class Text extends React.Component<Props> {
  static childContextTypes = {
    isNested: PropTypes.bool,
  };

  static contextTypes = {
    isNested: PropTypes.bool,
  };

  getChildContext = () => ({
    isNested: true,
  });

  render() {
    let style = this.props.style;

    if (this.context.isNested === undefined) {
      // default styling must be applied only on the root Text component
      // `undefined` style in nested Text components indicates style inheritance
      style = [styleSheet.defaultTextStyle, this.props.style];
    }

    return (
      // $FlowFixMe Errors after moving rn modules from untyped to declarations
      <ReactNative.Text {...this.props} style={style}>
        {this.props.children}
      </ReactNative.Text>
    );
  }
}

const styleSheet = StyleSheet.create({
  // These values are from the official design. Don't touch it please.
  defaultTextStyle: {
    fontWeight: 'normal',
    color: defaultTokens.colorTextPrimary,
    android: {
      fontSize: 15,
      letterSpacing: 0,
      fontFamily: 'Roboto-Regular',
    },
    ios: {
      fontSize: 14,
      letterSpacing: -0.15,
      fontFamily: 'CircularPro-Book',
    },
  },
});
