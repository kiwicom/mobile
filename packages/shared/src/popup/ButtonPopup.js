// @flow

import * as React from 'react';
import { View } from 'react-native';

import Popup from './Popup';
import TextButton from '../buttons/TextButton';
import StyleSheet from '../PlatformStyleSheet';
import CloseButton from '../buttons/CloseButton';
import type { TranslationType } from '../../types/Translation';

type Props = {|
  +buttonTitle: TranslationType,
  +children: React.Node,
  +isVisible: boolean,
  +onClose: Function,
  +onSave: Function,
  +style?: Object,
  +testID?: string,
|};

const POPUP_PADDING = 15;
const POPUP_BUTTONS_MARGIN = 10;

/**
 * Simple popup with confirmation button at the bottom.
 */
export default class ButtonPopup extends React.Component<Props> {
  onClose = () => this.props.onClose();

  onSave = () => this.props.onSave();

  render() {
    return (
      <Popup
        isVisible={this.props.isVisible}
        onClose={this.onClose}
        bottomContent={
          /* $FlowFixMe Errors after moving rn modules from untyped to
           * declarations */
          <View style={styles.buttonsWrapper}>
            {/* $FlowFixMe Errors after moving rn modules from untyped to
             * declarations */}
            <View style={[styles.button, styles.buttonFirst]}>
              <CloseButton onPress={this.onClose} />
            </View>
            {/* $FlowFixMe Errors after moving rn modules from untyped to
             * declarations */}
            <View style={styles.button}>
              <TextButton title={this.props.buttonTitle} onPress={this.onSave} type="primary" />
            </View>
          </View>
        }
      >
        {/* $FlowFixMe Errors after moving rn modules from untyped to
         * declarations */}
        <View style={[styles.content, this.props.style && this.props.style.content]}>
          {this.props.children}
        </View>
      </Popup>
    );
  }
}

const styles = StyleSheet.create({
  content: {
    padding: POPUP_PADDING,
    paddingBottom: 0,
  },
  buttonsWrapper: {
    flexDirection: 'row',
  },
  button: {
    margin: POPUP_BUTTONS_MARGIN,
    flex: 1,
  },
  buttonFirst: {
    marginEnd: 0,
  },
});
