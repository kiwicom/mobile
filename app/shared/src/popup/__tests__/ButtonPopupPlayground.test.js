// @flow

import * as React from 'react';
import { PlaygroundRenderer } from '@kiwicom/react-native-app-playground';
import { DummyTranslation } from '@kiwicom/react-native-app-translations';

import Text from '../../Text';
import ButtonPopup from '../ButtonPopup';

const noop = () => {};

describe('ButtonPopup playground', () => {
  it('renders in playground', () => {
    PlaygroundRenderer.render(
      <ButtonPopup
        buttonTitle={<DummyTranslation id="Save" />}
        isVisible={true}
        onSave={noop}
        onClose={noop}
      >
        <Text>Child</Text>
      </ButtonPopup>,
    );
  });
});
