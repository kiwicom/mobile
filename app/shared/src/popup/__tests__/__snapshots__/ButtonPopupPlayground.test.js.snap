// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonPopup playground renders in playground 1`] = `
<Popup
  isVisible={true}
  onClose={[Function]}
>
  <View
    style={
      Array [
        Object {
          "padding": 20,
          "paddingBottom": 0,
        },
        undefined,
      ]
    }
  >
    <Text>
      Child
    </Text>
  </View>
  <View
    style={
      Object {
        "margin": 20,
        "marginTop": 20,
      }
    }
  >
    <Button
      onPress={[Function]}
      styles={
        Object {
          "buttonWrapper": Object {
            "height": 44,
          },
        }
      }
      title={
        <DummyTranslation
          id="Save"
        />
      }
    />
  </View>
</Popup>
`;
